torch/csrc/distributed/rpc/tensorpipe_agent.cpp

```diff
@@ -0,0 +1,380 @@
+      requestMessage.isRequest(),
+      "TensorPipeAgent::send(..) is only for sending requests.");
+
+  if (!isRunning_) {
+    auto err = c10::str(
+        "Node ",
+        RpcAgent::getWorkerInfo().id_,
+        "tried to send() a message of type ",
+        requestMessage.type(),
+        " but RPC is no longer running on this node.");
+    throw std::runtime_error(err);
+  }
+
+  const auto& url = findWorkerURL(toWorkerInfo);
+
+  std::lock_guard<std::mutex> lock(mutex_);
+
+  // See if we already have a connection to this address or not
+  auto it = connectedPipes_.find(toWorkerInfo.id_);
+  if (it == connectedPipes_.end()) {
+    std::tie(it, std::ignore) = connectedPipes_.emplace(
+        toWorkerInfo.id_, context_->connect(url));
+  }
+  std::shared_ptr<tensorpipe::Pipe>& pipe = it->second;
+
+  std::shared_ptr<FutureMessage> futureResponseMessage =
+      std::make_shared<FutureMessage>();
+  requestMessage.setId(nextMessageID_++);
+  pendingResponseMessage_[requestMessage.id()] = futureResponseMessage;
+
+  pipeWrite(
+      pipe,
+      std::move(requestMessage),
+      [this, &pipe, futureResponseMessage](const tensorpipe::Error& error) {
+        if (error) {
+          LOG(WARNING) << "Write callback error: " << error.what();
+          futureResponseMessage->setError(error.what());
+          return;
+        }
+
+        pipeRead(
+            pipe,
+            [this](const tensorpipe::Error& error, Message&& responseMessage) {
+              // Identify future response message by message ID
+              uint64_t messageId = responseMessage.id();
+              std::shared_ptr<FutureMessage> futureResponseMessage;
+              {
+                std::lock_guard<std::mutex> lock(mutex_);
+                auto it = pendingResponseMessage_.find(messageId);
+                TORCH_INTERNAL_ASSERT(
+                    it != pendingResponseMessage_.end(),
+                    "message ID ",
+                    messageId,
+                    " is not recognized");
+                futureResponseMessage = std::move(it->second);
+                pendingResponseMessage_.erase(it);
+              }
+
+              if (error) {
+                LOG(WARNING) << "Read response error: " << error.what();
+                futureResponseMessage->setError(error.what());
+                return;
+              }
+
+              if (responseMessage.type() == MessageType::EXCEPTION) {
+                futureResponseMessage->setError(std::string(
+                    responseMessage.payload().begin(),
+                    responseMessage.payload().end()));
+              } else {
+                futureResponseMessage->markCompleted(
+                    std::move(responseMessage));
+              }
+            });
+      });
+
+  return futureResponseMessage;
+}
+
+// TODO: Remove sync()
+void TensorPipeAgent::sync() {}
+
+// TODO: Remove join()
+void TensorPipeAgent::join() {
+  shutdown();
+}
+
+void TensorPipeAgent::shutdown() {
```

Comment:
After https://github.com/pytorch/pytorch/pull/36239, this should now be `shutdownImpl`. The destructor and `join` can still call `shutdown`